sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -404,20 +404,25 @@ object ExtractSingleColumnNullAwareAntiJoin extends JoinSelectionHelper with Pre
    * which is very time consuming because it's an O(M*N) calculation.
    * But if it's a single column case O(M*N) calculation could be optimized into O(M)
    * using hash lookup instead of loop lookup.
+   *
+   * [SPARK-36665] IsNull(EqualTo(a=b)) gets optimized to Or(IsNull(a), IsNull(b))
+   * In addition, it is possible that Or(IsNull(a), IsNull(b)) gets optimized to IsNull(a)
+   * when a is nullable and b is not, vice versa
    */
   def unapply(join: Join): Option[ReturnType] = join match {
     case Join(left, right, LeftAnti,
-      Some(Or(e @ EqualTo(leftAttr: Expression, rightAttr: Expression),
-        IsNull(e2 @ EqualTo(_, _)))), _)
+      Some(Or(e @ EqualTo(leftAttr: Expression, rightAttr: Expression), IsNull(e2))), _)
```

Comment:
This means that `ExtractSingleColumnNullAwareAntiJoin ` may match more patterns. I'm really worried about correctness here, or we should have a more explicit way to mark null aware anti-join.

Reply:
I understand why you are worried. I reverted back the first case match in the next commit. I hope that makes you feel a bit more comfortable...

Reply:
I reverted it as I made it unaryExpressions only